### DISCUSSION TITLE
Add AAC_he and AAC_he_v2 support to new hls output feature (libfdk_aac)

### DISCUSSION THREAD

**Discussion by johnengstrom** [AzuraCast/AzuraCast#5502] [Ideas]
Hi,

The new HLS feature is amazing! I could not be more thrilled to see that added in the latest version and it's been working very well—no technical issues to report thus far (aside from that weird slim error during setup).

I do have one modest request. **_Would it be possible to add codec options for "aac_he" and  "aac_he_v2"?_**

I would like to change the codec on my 32kbps, 48kps and 64kbps streams from aac to aac_he and aac_he_v2 using libfdk_aac.

I find those two codecs sound MUCH nicer at the lower bitrates so I'm keen on changing that value ASAP.

Going through the source code it *seems* as if libfdk_aac is installed, so in theory codecs from that library could be used.  

Looking at \src\Radio\Backend\Liquidsoap\ConfigWriter.php, at line 1121, there is this block that matches what I see in the Liquidsoap configuration.  

              %audio(
                    codec="aac", <---- ADD SUPPORT HERE FOR aac_he AND aac_he_v2
                    samplerate=44100,
                    channels=2,
                    b="{$streamBitrate}k",
                    profile="aac_low"
                )
            

I tried copying the block from the liquidsoap file to the textarea below that block and changing the value but it did not seem to override the generated code (was a long-shot).  It seems the source code needs a change.

If it's possible to override in the setting in the textarea, that would be terrific!

Thanks!

John

PS:  Handy tools for validating your HLS streams are Apple's mediastreamvalidator and hlsreport.  [https://developer.apple.com/documentation/http_live_streaming/using_apple_s_http_live_streaming_hls_tools](https://developer.apple.com/documentation/http_live_streaming/using_apple_s_http_live_streaming_hls_tools)

**Comment by Vaalyn** [comment]
I have been experimenting with switching to `fdk_aac` but I'm running into an issue with FFmpeg:

```
The requested ffmpeg encoder was not found, please make sure that ffmpeg was compiled with support for it
```

It looks like the package repositories for Ubuntu (and other distros) don't include the `fdk_aac` encoder in their builds of FFmpeg so we'd have to manually compile FFmpeg with support for it in our build process.

**Reply by johnengstrom** [reply]
Indeed.  To enable these codecs you must compile `ffmpeg` with `libfdk_aac` library included.  Not a difficult task but—but not trivial.  

I wonder if a third-party library like `libfdk_aac` can be added to the `ffmpeg` part of the installation script.  

Another option would be adding an option to upload a platform compatible `ffmpeg` binary, like other features/code that can't be included in the git repo.

I also am curious to know if it's possible to see the final `ffmpeg` output string generated by Azuracast.  I'll dig into the source code and see if I can answer my own question :)

**Comment by johnengstrom** [comment]
Looking at the Ansible playbook it seems like all it does is install the standard distro version of ffmpeg (that does not include  `libfdk_aac`).   Thinking an option to upload your own pre-compiled binary and the ability to change the codec string would be needed to make this work.

**Comment by Rafi594** [comment]
Hello !

I managed to add the support for AAC_HE by adding this PPA inside the AzuraCast container https://launchpad.net/~sergey-dryabzhinsky/+archive/ubuntu/ffmpeg, reinstalling ffmpeg and replacing the part for HLS in the Liquidsoap configuration with this :
```

aac_low = %ffmpeg(
    format="mpegts",
    %audio(
        codec="libfdk_aac",
        samplerate=44100,
        channels=2,
        b="48k",
        profile="aac_he"
    )
)
```

The only issue with this workaround is the PPA will be removed with the updates.
